src/main/java/org/myrobotlab/codec/CodecUtils.java

```diff
@@ -34,51 +24,210 @@
 import org.yaml.snakeyaml.Yaml;
 import org.yaml.snakeyaml.constructor.Constructor;
 
-import com.google.gson.Gson;
-import com.google.gson.GsonBuilder;
-import com.google.gson.internal.LinkedTreeMap;
+import java.io.*;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.stream.Collectors;
 
 /**
  * handles all encoding and decoding of MRL messages or api(s) assumed context -
  * services can add an assumed context as a prefix
  * /api/returnEncoding/inputEncoding/service/method/param1/param2/ ...
- * 
+ *
  * xmpp for example assumes (/api/string/gson)/service/method/param1/param2/ ...
- * 
+ *
  * scheme = alpha *( alpha | digit | "+" | "-" | "." ) Components of all URIs: [
  * &lt;scheme&gt;:]&lt;scheme-specific-part&gt;[#&lt;fragment&gt;]
- * http://stackoverflow.com/questions/3641722/valid-characters-for-uri-schemes
- * 
+ *
  * branch API test 5
+ *
+ * @see <a href="http://stackoverflow.com/questions/3641722/valid-characters-for-uri-schemes">Valid characters for URI schemes</a>
  */
 public class CodecUtils {
 
   public final static Logger log = LoggerFactory.getLogger(CodecUtils.class);
 
+  /**
+   * A description of an API type
+   */
   public static class ApiDescription {
-    String key;
-    String path; // {scheme}://{host}:{port}/api/messages
-    String exampleUri;
-    String description;
+    /**
+     * The string to use after {@link #PARAMETER_API}
+     * in URIs to select this API.
+     */
+    public final String key;
 
+    /**
+     * The path to reach this API
+     */
+    public final String path; // {scheme}://{host}:{port}/api/messages
+
+    /**
+     * An example URI to reach this API
+     */
+    public final String exampleUri;
+
+    /**
+     * The description of this API
+     */
+    public final String description;
+
+    /**
+     * Construct a new API description.
+     *
+     * @param key {@link #key}
+     * @param uriDescription {@link #path}
+     * @param exampleUri {@link #exampleUri}
+     * @param description {@link #description}
+     */
     public ApiDescription(String key, String uriDescription, String exampleUri, String description) {
       this.key = key;
       this.path = uriDescription;
       this.exampleUri = exampleUri;
       this.description = description;
     }
+

```

Comment:
equals and hashCode for a utils ?

Reply:
From what I understand, ApiDescription is a dataclass whose only purpose is to hold information, it's generally considered required to add equals and hashcode for dataclasses so if we ever wanted to check for equality or put them in a map it won't rely on the identity of the object. It is more boilerplate though, I don't believe we're actively using equals and hashcode except in tests so if we need to they can be removed